tor: respect non-onion -onlynet= for outgoing Tor connections

vasild opened the pull request:
If
* `-onlynet=` is given one or more times but none of them includes
  `onion` and
* `-proxy` is not set and
* `-onion` is not set,

then we should not be making outbound connections to the Tor network.

Fixes https://github.com/bitcoin/bitcoin/issues/22647

vasild committed:
tor: respect non-onion -onlynet= for outgoing Tor connections

If
* `-onlynet=` is given one or more times but none of them includes
  `onion` and
* `-proxy` is not set and
* `-onion` is not set,
then we should not be making outbound connections to the Tor network.

Fixes https://github.com/bitcoin/bitcoin/issues/22647

vasild commented:
See also https://en.wikipedia.org/wiki/Principle_of_least_astonishment

jonatack commented:
Concept ACK. It may be good to make it a separate function with test coverage.

jonatack commented on src/torcontrol.cpp:
```diff
@@ -373,9 +373,18 @@ void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply&
+                }
+            }
+        }
+        if (gArgs.GetArg("-onion", "") == "" && allow_outbound_onion) {
```
A function seems nicer than a localvar here (optionally declare in header file for unit testing)
```diff
+bool AllowOutboundOnion()
+{
+    if (gArgs.IsArgSet("-onlynet")) return false;
+    for (const auto& net : gArgs.GetArgs("-onlynet")) {
+        if (ParseNetwork(net) == NET_ONION) return true;
+    return false;
+}
+
 void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply& reply)
 {
     if (reply.code == 250) {
         LogPrint(BCLog::TOR, "tor: Authentication successful\n");
 
         // Now that we know Tor is running, possibly set the proxy for onion addresses.
-        bool allow_outbound_onion{true};
-        if (gArgs.IsArgSet("-onlynet")) {
-            allow_outbound_onion = false;
-            for (const auto& net : gArgs.GetArgs("-onlynet")) {
-                if (ParseNetwork(net) == NET_ONION) {
-                    allow_outbound_onion = true;
-                    break;
-                }
-            }
-        }
-        if (gArgs.GetArg("-onion", "") == "" && allow_outbound_onion) {
+        if (gArgs.GetArg("-onion", "").empty() && AllowOutboundOnion()) {
             CService resolved(LookupNumeric("127.0.0.1", 9050));
             proxyType addrOnion = proxyType(resolved, true);
```

jonatack replied:
Tested with this patch on mainnet:
- `onlynet=i2p`: only I2P outbound connections are made (i2pd 2.23)
- `onlynet=ipv4`: only IPv4 outbound connections are made
- `onlynet=onion`: only onion outbound connections are made (tor 0.4.6)

jonatack replied:
Tested combinations (all tests using the proposed function version ;)
- `onlynet=i2p` and `onlynet=onion`: only I2P and onion outbound connections are made
- `onlynet=ipv4` and `onlynet=onion`: only IPv4 and onion outbound connections are made

(Test coverage for these cases--single onlynet and combinations--would be great.)

prayank23 replied:
> Test coverage for these cases--single onlynet and combinations--would be great

Agree

vasild replied:
I think it is fine either way. A function would make more sense if this is used in more than one place. The code is now covered by a new functional test `feature_onlynet.py`. Left it as is.

jonatack replied:
I think the function is not only shorter but also clearer and easier to reason about.

prayank23 commented:
> See also https://en.wikipedia.org/wiki/Principle_of_least_astonishment

This link says "The behavior should not astonish or surprise users"

I am surprised by lot of things in Bitcoin Core 😄

Concept ACK. Doesn't make sense to create outbound connections with onion peers if `onlynet=i2p` is used.

prayank23 commented on src/torcontrol.cpp:
```diff
@@ -373,9 +373,18 @@ void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply&
+            }
+        }
+        if (gArgs.GetArg("-onion", "") == "" && allow_outbound_onion) {
             CService resolved(LookupNumeric("127.0.0.1", 9050));
```
This is normally 9151 on Windows so can we use #ifdef here?

```suggestion
#ifdef WIN32 
            CService resolved(LookupNumeric("127.0.0.1", 9150));
#else
            CService resolved(LookupNumeric("127.0.0.1", 9050));
#endif
```

vasild replied:
Yes, ouf of scope of this PR. This belongs to https://github.com/bitcoin/bitcoin/pull/15423, I hope that PR moves forward.

prayank23 commented:
> It may be good to have functional test coverage (or make it a separate function with unit tests).

Agree. Do you think these steps are correct for test?

1. Run Node 1: 
```
bitcoind -port=18333 -rpcport=18222 -datadir="/home/user/node1" -regtest=1 -listen=1 -server=1 -debug=net -rpcuser=user1 -rpcpassword=password1 -torcontrol='127.0.0.1:9051' -proxy='127.0.0.1:9050' -onlynet=onion
```
2. Run Node 2:
```
bitcoind -port=18777 -rpcport=18666 -datadir="/home/user/node2" -regtest=1 -listen=1 -server=1 -debug=net -rpcuser=user2 -rpcpassword=password2 -addnode='127.0.0.1:18333' -torcontrol='127.0.0.1:9051' -proxy='127.0.0.1:9050' -onlynet=onion
```
3. Stop Node 1 and 2:

```
bitcoin-cli -rpcport=18222 -rpcuser=user1 -rpcpassword=password1 stop
bitcoin-cli -rpcport=18666 -rpcuser=user2 -rpcpassword=password2 stop
```
4. Restart both Node 2 with different config (`onlynet=i2p` and no `proxy`):

```
bitcoind -port=18777 -rpcport=18666 -datadir="/home/user/node2" -regtest=1 -listen=1 -server=1 -debug=net -rpcuser=user2 -rpcpassword=password2 -i2psam=127.0.0.1:7656 -onlynet=i2p
```
5. Check debug.log:

```
trying connection 3llpwlxkisrm2iu5oayh2hd6df7q36wdmt6nounenzqcxkasxfk34eid.onion
```
Master branch should print such connection attempts and PR branch should not

jonatack commented:
Checking the debug log output is a bit of a last resort when there's nothing else to assert on... getnetworkinfo limited/reachable (and maybe getpeerinfo) would probably be good.

wodry commented:
Would this fix also #13378 ?

prayank23 commented:
> Checking the debug log output is a bit of a last resort when there's nothing else to assert on... getnetworkinfo limited/reachable (and maybe getpeerinfo) would probably be good.

Also the steps I mentioned above don't work. They worked on one of my VMs. Maybe it already had some onion address in peers.dat. I was expecting it will add onion peer in peers.dat but it didn't. `addpeeraddress` with some onion address and port followed by `getpeerinfo` should work.

prayank23 commented:
Wrote a PowerShell [script ](https://github.com/prayank23/bitcoin-ps-scripts/blob/main/Scripts/Node/onlynet_i2p.ps1) to test this. Tried on Windows 10 and Pop!_OS. Node 2 is connected to Node 1 on Linux and fails on Windows.

TL;DR
   1. Run Node 1 and get onion URL for it. 
   2. Sleep 10 seconds
   3. Get onion address for Node 1 from `getnetworkinfo` -> localaddresses -> address ending with .onion and its port
   4. Run Node 2 with `onlynet=i2p` (No Tor proxy)
   5. Sleep 10 seconds
   6.  Add Node 1 in peers.dat for Node 2 using hidden RPC `addpeeraddress`
   7. Sleep 30 seconds
   8. Check if Node 2 is connected to Node 1 with `getpeerinfo`

```
Peer info:

xejoddwvi35krze3546qtx6dfmednjs2ccsociyezkyhzdxtemp5v5ad.onion:18444
outbound-full-relay
```

vasild commented:
> Would this fix also #13378 ?

To my understanding, yes, but only if none of `-proxy` or `-onion` is set.

Providing both `-onlynet=ipv4 -onion=127.0.0.1:9050` is kind of contradictory - why provide an outgoing tor proxy if you only want to connect to IPv4? I think Bitcoin Core should print an error and refuse to start in this case, but instead it treats it as "onlynet ipv4 or tor". Changing/fixing that is out of the scope of this PR which only aims to fix the behavior if none of `-proxy` or `-onion` is set.

wodry commented:
Also of course out of this PR, just to mention: I am sure you are already aware that to have multiple "only" nets is paradox, the option should better be called e.g. "net" instead of "onlynet"..

vasild committed:
test: ensure -onlynet behaves as expected

Introduce a basic TCP server in the functional testing framework and use
it to simulate a Tor Control daemon, which is needed in order for the
code in `src/torcontrol.cpp` to execute the code that is relevant for
`-onlynet`.

vasild commented:
`4f4c7d8b11..4f3020d524`: append a functional test to exercise the modified behavior

vasild commented:
`4f3020d524...00b7ba51eb`: hush `test/lint/lint-python.sh`

vasild commented:
`00b7ba51eb...25f9e113db`: change `allow_outbound_onion` to a function, as per suggestion

jonatack commented on test/functional/feature_onlynet.py:
```diff
@@ -0,0 +1,54 @@
+
+class OnlynetTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.extra_args = [
```
I needed this line for the test to run:
```diff
 class OnlynetTest(BitcoinTestFramework):
     def set_test_params(self):
+        self.setup_clean_chain = True
         self.extra_args = [
```
Otherwise, was seeing the following error:

<details><summary>test output</summary><p>

```
$ test/functional/feature_onlynet.py 
2021-08-11T09:22:40.305000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_p1tdzv8m
2021-08-11T09:22:41.628000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 130, in main
    self.setup()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 285, in setup
    self.setup_network()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 379, in setup_network
    self.setup_nodes()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 409, in setup_nodes
    assert_equal(n.getblockchaininfo()["blocks"], 199)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/util.py", line 49, in assert_equal
    raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
AssertionError: not(0 == 199)
2021-08-11T09:22:41.680000Z TestFramework (INFO): Stopping nodes
2021-08-11T09:22:41.860000Z TestFramework (WARNING): Not cleaning up dir /tmp/bitcoin_func_test_p1tdzv8m
2021-08-11T09:22:41.860000Z TestFramework (ERROR): Test failed. Test logging available at /tmp/bitcoin_func_test_p1tdzv8m/test_framework.log
2021-08-11T09:22:41.861000Z TestFramework (ERROR): 
2021-08-11T09:22:41.861000Z TestFramework (ERROR): Hint: Call /home/jon/projects/bitcoin/bitcoin/test/functional/combine_logs.py '/tmp/bitcoin_func_test_p1tdzv8m' to consolidate all logs
2021-08-11T09:22:41.862000Z TestFramework (ERROR): 
2021-08-11T09:22:41.862000Z TestFramework (ERROR): If this failure happened unexpectedly or intermittently, please file a bug and provide a link or upload of the combined log.
2021-08-11T09:22:41.862000Z TestFramework (ERROR): https://github.com/bitcoin/bitcoin/issues
2021-08-11T09:22:41.862000Z TestFramework (ERROR): 
```
</p></details>

jonatack replied:
Hm, after coming back here and rebuilding, the test now runs for me without it. Strange.  Sorry for the noise.

jonatack commented on test/functional/feature_onlynet.py:
```diff
@@ -0,0 +1,54 @@
+        # Passes with or without https://github.com/bitcoin/bitcoin/pull/22651,
+        # just to ensure behavior does not change unintended.
+        n2 = networks_dict(self.nodes[2].getnetworkinfo())
+        assert_equal(n2['onion']['reachable'], True)
```
Here is a proposed simplification using a dict comprehension in `networks_dict` and a loop for the assertions:
```diff
+    def networks_dict(self, node):
+        info = self.nodes[node].getnetworkinfo()
+        return {network["name"]: network for network in info['networks']}
 
     def run_test(self):
-        def networks_dict(d):
-            r = {}
-            for x in d['networks']:
-                r[x['name']] = x
-            return r
-
-        # Would fail without https://github.com/bitcoin/bitcoin/pull/22651
-        n0 = networks_dict(self.nodes[0].getnetworkinfo())
-        assert_equal(n0['onion']['reachable'], False)
-
-        # Passes with or without https://github.com/bitcoin/bitcoin/pull/22651,
-        # just to ensure behavior does not change unintended.
-        n1 = networks_dict(self.nodes[1].getnetworkinfo())
-        assert_equal(n1['onion']['reachable'], True)
-
-        # Passes with or without https://github.com/bitcoin/bitcoin/pull/22651,
-        # just to ensure behavior does not change unintended.
-        n2 = networks_dict(self.nodes[2].getnetworkinfo())
-        assert_equal(n2['onion']['reachable'], True)
+        for node, reachable in [[0, False], [1, True], [2, True]]:
+            assert_equal(self.networks_dict(node)['onion']['reachable'], reachable)
```
Could perhaps add a comment describing why we expect the first node to not be reachable.

vasild replied:
Done something similar. I find `return {network["name"]: network for network in info["networks"]}` extremely hard to read.

jonatack replied:
Comprehensions seem idiomatic/pythonic and the most-used way to do it in this codebase too (though I agree that a shorter `n` or `net` for the one line would probably be better than `network`), but I like your new version too. Nice and simple.

jonatack commented on test/functional/test_framework/basic_server.py:
```diff
@@ -0,0 +1,93 @@
+        """
+        # create_server() is only in python 3.8
+        # https://docs.python.org/3/library/socket.html#socket.create_server
+        #self.listen_socket = socket.create_server(address=bind, reuse_port=True)
```
```suggestion
        # self.listen_socket = socket.create_server(address=bind, reuse_port=True)
```

vasild replied:
Done.

jonatack commented on test/functional/test_framework/basic_server.py:
```diff
@@ -0,0 +1,93 @@
+        self.listen_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        self.listen_socket.bind(bind)
+        self.listen_socket.listen()
+        #
```
?

vasild replied:
I wanted to mark the end of the snippet that is equivalent to `create_server()`. Removed.

jonatack reviewed (commented):
Nice!  A few suggestions for the test.  Only skimmed the basic server so far.

jonatack commented on test/functional/feature_onlynet.py:
```diff
@@ -0,0 +1,54 @@
+
+from test_framework.basic_server import (
+    BasicServer,
+    tor_control
```
nit, either add a comma at the end of this line, or one-line it:
```python
from test_framework.basic_server import BasicServer, tor_control
```

vasild commented:
`test_framework/basic_server.py` can also be used to add some functional tests for I2P. We can simulate an I2P router at least as long as the text/line-based SAM is being talked on the socket. Later, when the socket switches from SAM to the binary bitcoin p2p protocol we would have to close the socket.

vasild commented on test/functional/feature_onlynet.py:
```diff
@@ -0,0 +1,54 @@
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""
+Test -onlynet
```
Done.

jonatack commented on test/functional/feature_onlynet.py:
```diff
@@ -0,0 +1,48 @@
+    def run_test(self):
+        # Node 0 would fail the check without https://github.com/bitcoin/bitcoin/pull/22651.
+        # Nodes 1 and 2 pass with or without https://github.com/bitcoin/bitcoin/pull/22651,
+        # they are here just to ensure behavior does not change unintended.
```
nits:

- I think it's more robust to describe why than to refer to a GitHub PR or depend on GitHub in general

- nit s/unintended/unintentionally/

vasild replied:
I agree, removed the github reference.

I think "uninternationally" is even better, changed.

jonatack commented on src/torcontrol.cpp:
```diff
@@ -368,14 +368,26 @@ void TorController::add_onion_cb(TorControlConnection& _conn, const TorControlRe
-        // if -onion isn't set to something else.
-        if (gArgs.GetArg("-onion", "") == "") {
+        // Now that we know Tor is running, possibly set the proxy for onion addresses.
+        if (gArgs.GetArg("-onion", "") == "" && AllowOutboundOnion()) {
```
```suggestion
        if (gArgs.GetArg("-onion", "").empty() && AllowOutboundOnion()) {
```

vasild replied:
You suggested this one earlier and I deliberately skipped it but did not explain why.

Here we are testing whether the result of `GetArg()` equals to its second argument, so it is better to keep that obvious, e.g. `GetArg(..., X) == X`.

Talkless replied:
Because `==` will generate/call string comparison code, instead of just checking string length using `empty()`. Better just use specialized member functions instead of some kind "generic" comparison.

Another example is using `clear()` vs `= ""`: https://www.youtube.com/watch?v=3X9qK7HWxjk

vasild replied:
I think readability trumps micro-optimization here, leaving it as is.

jonatack commented on test/functional/feature_onlynet.py:
```diff
@@ -0,0 +1,48 @@
+            ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1'],
+            ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1', '-onion=127.0.0.1:9050'],
+            ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1', '-proxy=127.0.0.1:9050'],
+        ]
```
Would it make sense to test with multiple onlynet settings?

```diff
diff --git a/test/functional/feature_onlynet.py b/test/functional/feature_onlynet.py
index 60e4caf33d..c43a541196 100755
--- a/test/functional/feature_onlynet.py
+++ b/test/functional/feature_onlynet.py
@@ -22,6 +22,8 @@ class OnlynetTest(BitcoinTestFramework):
             ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1'],
             ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1', '-onion=127.0.0.1:9050'],
             ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1', '-proxy=127.0.0.1:9050'],
+            ['-onlynet=ipv4', '-onlynet=ipv6', '-torcontrol=127.0.0.1:19051', '-listenonion=1'],
         ]
         self.num_nodes = len(self.extra_args)

@@ -39,7 +41,7 @@ class OnlynetTest(BitcoinTestFramework):
-        for node, expected_reachable in [[0, False], [1, True], [2, True]]:
+        for node, expected_reachable in [[0, False], [1, True], [2, True], [3, False]]:
             self.log.info(f'Verifying node {node} which uses {self.extra_args[node]}')
             assert_equal(self.onion_is_reachable(node), expected_reachable)
```

vasild replied:
I think it is fine as it is. The point is to have `onlynet=...` but not `onlynet=onion`.

jonatack reviewed (commented):
ACK 16bf93e1361fa9a1503cadd96e5121d92937bfb5 rebased to master, debug build, ran new tests, verified that the test fails where expected on master, and previously manually verified the behavior extensively as described in https://github.com/bitcoin/bitcoin/pull/22651#discussion_r684429929 and https://github.com/bitcoin/bitcoin/pull/22651#discussion_r684434758

Happy to re-ACK for further improvements.

Talkless commented:
Concept ACK

prayank23 reviewed (approved):
tACK https://github.com/bitcoin/bitcoin/pull/22651/commits/16bf93e1361fa9a1503cadd96e5121d92937bfb5

Tested with [test/functional/feature_onlynet.py](https://raw.githubusercontent.com/bitcoin/bitcoin/16bf93e1361fa9a1503cadd96e5121d92937bfb5/test/functional/feature_onlynet.py) and my [PowerShell Script](https://github.com/bitcoin/bitcoin/pull/22651#issuecomment-894761007)

Master Branch:

1. Python test fails
```
2021-08-11T22:47:24.271000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_tfz0x_fp
2021-08-11T22:47:26.786000Z TestFramework (INFO): Verifying node 0 which uses ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1']
2021-08-11T22:47:26.790000Z TestFramework (ERROR): Assertion failed
```
2. PS script returns : `xejoddwvi35krze3546qtx6dfmednjs2ccsociyezkyhzdxtemp5v5ad.onion:18444 outbound-full-relay`

PR Branch:

1. Python test successful:
```
11T22:44:25.360000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_6vo22aqz
2021-08-11T22:44:26.622000Z TestFramework (INFO): Verifying node 0 which uses ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1']
2021-08-11T22:44:26.623000Z TestFramework (INFO): Verifying node 1 which uses ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1', '-onion=127.0.0.1:9050']
2021-08-11T22:44:26.625000Z TestFramework (INFO): Verifying node 2 which uses ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1', '-proxy=127.0.0.1:9050']
2021-08-11T22:44:26.629000Z TestFramework (INFO): Stopping nodes
2021-08-11T22:44:26.791000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_6vo22aqz on exit
2021-08-11T22:44:26.791000Z TestFramework (INFO): Tests successful
    
```
2. PS script returns nothing (no onion peer)

Although both tests use different approach and a basic TCP server is added in test framework for this test, most important commit is https://github.com/bitcoin/bitcoin/pull/22651/commits/c591379eb03381145c85ee5760ce829be337a749 which looks good and fixes the issue.

Thanks for fixing this issue and improving privacy in Bitcoin Core. Hope this is merged soon and tests can always be improved in future if required.

vasild commented:
`16bf93e136...5959ece29e`: address minor suggestions and avoid possibility of colliding listening ports by binding to port 0 (let the OS choose an available port, thanks @darosior for the idea!)

jonatack commented on test/functional/feature_onlynet.py:
```diff
@@ -0,0 +1,47 @@
+    def run_test(self):
+        # Node 0 would fail the check without c591379eb03381145c85ee5760ce829be337a749.
+        # Nodes 1 and 2 pass with or without that commit, they are here just to ensure
+        # behavior does not change unintentionally.
```
maybe use a docstring and describe the commit, feel free to ignore
```diff
-        # Node 0 would fail the check without c591379eb03381145c85ee5760ce829be337a749.
-        # Nodes 1 and 2 pass with or without that commit, they are here just to ensure
-        # behavior does not change unintentionally.
+        """
+        - The test of node 0 would fail without commit c591379eb0,
+          "tor: respect non-onion -onlynet= for outgoing Tor connections."
+
+        - The tests of nodes 1 and 2 pass with or without that commit;
+          they ensure that behavior does not change unintentionally.
+        """
```

vasild replied:
Leaving it as is.

jonatack commented on test/functional/feature_onlynet.py:
```diff
@@ -0,0 +1,47 @@
+        # Nodes 1 and 2 pass with or without that commit, they are here just to ensure
+        # behavior does not change unintentionally.
+        for node, expected_reachable in [[0, False], [1, True], [2, True]]:
+            self.log.info(f'Verifying node {node} which uses {self.extra_args[node]}')
```
output is fairly long, could shorten without loss of signal
```suggestion
            self.log.info(f'Test node {node} {self.extra_args[node]}')
```

vasild replied:
Done.

jonatack commented on test/functional/test_framework/basic_server.py:
```diff
@@ -0,0 +1,94 @@
+    def __init__(self, *, bind, response_generator):
+        """
+        :param bind: Listen address-port tuple, for example ("127.0.0.1", 80).
+        If the port is 0 then a random available one will be chosen and it
```
indeed, seeing a different random ports as described here

jonatack reviewed (commented):
ACK 5959ece29e5b6b7ce0d8be60ae64a3312861f841 tested rebased to master, verified that the test fails where expected on master, and previously manually verified the behavior extensively as described in https://github.com/bitcoin/bitcoin/pull/22651#discussion_r684429929 and https://github.com/bitcoin/bitcoin/pull/22651#discussion_r684434758

Talkless commented on src/torcontrol.cpp:
```diff
@@ -368,14 +368,26 @@ void TorController::add_onion_cb(TorControlConnection& _conn, const TorControlRe
+    if (!gArgs.IsArgSet("-onlynet")) {
+        return true;
+    }
+    for (const auto& net : gArgs.GetArgs("-onlynet")) {
```
Please consider using standard algorithms instead of hand-rolled loop:
```
#include <algorithm>
...

    const auto only_nets{gArgs.GetArgs("-onlynet")};

    //TODO: C++20: use std::ranges::any_of() instead
    return std::any_of(only_nets.cbegin(), only_nets.cend(), [](const auto &net) {
        return ParseNetwork(net) == NET_ONION;
    });
```

vasild replied:
Done.

jonatack commented:
This has a couple of ACKs and would be very good to have in 22.0 to alleviate user confusion over an impactful user-facing change in the release. @vasild, do you plan to update? Anyone else like to review here?

vasild commented:
`5959ece29e...a2ebcda19a`: address some suggestions

> @vasild, do you plan to update?

Done! :)

jonatack commented:
Code review re-ACK a2ebcda19a25067cb8183f25f6ffd0a26065fcc5 per `git diff 5959ece a2ebcda` plus sanity check of rebase on master, debug build, re-ran test a few times

Rspigler commented:
Concept ACK

prayank23 commented:
reACK https://github.com/bitcoin/bitcoin/pull/22651/commits/a2ebcda19a25067cb8183f25f6ffd0a26065fcc5

Major changes since last review:

1. https://github.com/bitcoin/bitcoin/pull/22651#discussion_r689088453
2. https://github.com/bitcoin/bitcoin/pull/22651#issuecomment-897498176

vasild commented on src/torcontrol.cpp:
```diff
@@ -368,14 +369,24 @@ void TorController::add_onion_cb(TorControlConnection& _conn, const TorControlRe
+    if (!gArgs.IsArgSet("-onlynet")) {
+        return true;
+    }
+    const auto& onlynets = gArgs.GetArgs("-onlynet");
```
Removed.

vasild commented:
`a2ebcda19a...baa6cb12ac`: address minor suggestion and fix commit id in a comment.

vasild commented:
@laanwj, I agree with your comments at https://github.com/bitcoin/bitcoin/pull/22648#discussion_r696535966. Opened https://github.com/bitcoin/bitcoin/pull/22834 which if merged will make this PR unnecessary.

jonatack commented:
Code review re-ACK baa6cb12ac5071bb68eb5ec375be73df4f4ae28f per `git diff a2ebcda baa6cb1`, rebase on master, debug build, ran the new test `test/functional/feature_onlynet.py` a few times, previously manually verified the behavior extensively as described in https://github.com/bitcoin/bitcoin/pull/22651#discussion_r684429929 and https://github.com/bitcoin/bitcoin/pull/22651#discussion_r684434758

Will review #22834 soon. Perhaps the functional test work here can be ~ported over to it if that change is preferred.

luke-jr commented:
#22834 looks like a better solution IMO

Talkless commented:
> #22834 looks like a better solution IMO

So this PR should be closed?

@vasild I'm confused :)

vasild commented:
Closing this in favor of https://github.com/bitcoin/bitcoin/pull/22834 which has more (concept) ACKs.

Thanks, @Talkless!
